pkg/trace/api/api.go

```diff
@@ -258,19 +271,10 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
 				msg = msg[:150] + "... (set DEBUG log level for more info)"
 			}
 			log.Errorf(msg)
-		} else {
-			select {
-			case r.Out <- trace:
-				// if our downstream consumer is slow, we drop the trace on the floor
-				// this is a safety net against us using too much memory
-				// when clients flood us
-			default:
-				atomic.AddInt64(&ts.TracesDropped, 1)
-				atomic.AddInt64(&ts.SpansDropped, int64(spans))
-
-				log.Errorf("dropping trace; reason: rate-limited")
-			}
+			continue
 		}
+
+		r.Out <- trace
```

Comment:
This seems like a pretty big change that we should be careful with. 

The current behavior of the agent under high load (when it can't send out traces as fast as it's getting them) is to just drop traces. So the response time for `/v0.2/traces` requests is always very fast, regardless of whether we're dropping or accepting traces. 

Now when the agent is under high load the response time for `/v0.2/traces` could be very high, meaning we'll just push the queuing up into the client integrations via pending HTTP requests. So we'll have effectively two queues to reason about: the go channel, and the pending http requests. 

If with this we'll just end up moving the dropping responsibility from the agent into the client integrations, would it be better to go back to non-blocking channel writes and simply reject the requests when the queue is full instead of dropping them? 

Also, by doing blocking here might we also hit the agent's connection limit if there's too many pending requests and drown out requests to other endpoints? 

If we really want to use blocking here, then we should probably add metrics to track the number of pending http requests (as well as the number of requests rejected by the server due to hitting the connection limit) so we can know when the agent is overloaded and not able to keep up.

Reply:
@gbbr for your review. I think we should check all these points.

Reply:
Oh, my. It seems you're right. I have been misleading myself and other reviewers (you're not the first to pose this question) to think that the client closes of the connection early because we reply [here](https://github.com/DataDog/datadog-agent/blob/d3dbe9da0b48b36f6b5d3694df221c7cd11ca40d/pkg/trace/api/api.go#L238), long before starting to process things.

I have manually checked this and the above assumption is wrong. The only way to finish a client request in the Go standard library is to [return from the handler](https://github.com/golang/go/blob/2c802e9/src/net/http/server.go#L1876-L1881). The current state would have indeed stalled connections.

The only solution I can think of here is to move the for loop into its own goroutine. This will not only solve the above mentioned problem but will also create a much lower latency handler, at a low cost.

Reply:
Looks good now. So with this change we basically get a new configurable buffer (via max memory setting) in the form of pending goroutines blocked on writing to the channel, and all http requests are still served fast, non-blocking.